### DISCUSSION TITLE
How to include a figure in PDF format?

### DISCUSSION THREAD

**Discussion by roaldarbol** [quarto-dev/quarto-cli#11361] [Q&A] [answered]
### Description

When including a PDF as an image, which is quite common for academic papers, it renders as a scroll box with little height. The basic example could look like this (I have also attempted to add e.g. `{height=100%}` to no avail):
```
:::{#fig-myfigure}

![](figure-1.pdf)

My caption
:::
```
...which renders like this:
<img width="781" alt="Screenshot 2024-11-11 at 13 28 00" src="https://github.com/user-attachments/assets/75666d02-704b-4047-adb8-ba51deef0899">

I guess the underlying issue is that it is inserted as in a PDF reader; I would expect it to render as a normal figure when it is inside a `#fig` div. Is there anyway to make it so?
Thanks in advance!

**Comment by mcanouil** [comment]
Please provide a fully reproducible example not a snippet out of context.

I will guess, the format is HTML.
I'm afraid a PDF is not an image, it's a binary format, thus it requires third party to be read which leads Pandoc to use iframe in HTML.
That's beyond Quarto or Pandoc control, that's how HTML works.

**Reply by roaldarbol** [reply]
It is indeed HTML. MyST converts PDFs to images with `imagemagick` before rendering, so I imagined a similar functionality to be present in Quarto, since many journals recommend submitting figures in PDF (e.g. [nature](https://research-figure-guide.nature.com/figures/preparing-figures-our-specifications/#exporting), [science](https://www.science.org/content/page/instructions-preparing-revised-manuscript), [journal of experimental biology](https://journals.biologists.com/jeb/pages/manuscript-prep#4.)) hence the question. Thanks for the answer. :-)

**Reply by mcanouil** [reply]
Quarto does not depend on `imagemagick` and will very unlikely do so.
You can make a Lua filter to do that if you want.

PS: I am perfectly aware of journals guidelines (I spent more than 10 years in academia).

**Reply by roaldarbol** [reply]
Oh, that wasn't meant to sound like I questioned you - apologies of it sounded that way; it was just to ensure context, just as you explained PDF format although I've also spent 10 years in academia and know what it is - doesn't mean I didn't think it was relevant information. :-)

Fair game regarding the `imagemagick` dependency, I may have a look at how to make a Lua filter then.

**Reply by mcanouil** [reply]
> Oh, that wasn't meant to sound like I questioned you - apologies of it sounded that way; it was just to ensure context, just as you explained PDF format although I've also spent 10 years in academia and know what it is - doesn't mean I didn't think it was relevant information. :-)

All is fine, I was only giving context on my background ;)

> Fair game regarding the imagemagick dependency, I may have a look at how to make a Lua filter then.

See https://github.com/search?q=lua%20imagemagick&type=repositories to get you started.

https://quarto.org/docs/extensions/filters.html

**Reply by roaldarbol** [reply]
Thanks for the links! I'll report back if I get around to making one. ;)

**Comment by roaldarbol** [comment, marked as answer]
Ended up creating a Lua filter which converts PDF files to `.webp` format with some help from ChatGPT, which does this quite nicely. You can tweak the `imagemagick` settings a to your liking. 

`pdf-to-webp.lua`, which I've for now just placed in the base of my workspace:
1. Converts `.pdf` to `.webp` in the source folder.
2. Except if a `webp` file is already generated
3. Checks whether the source `.pdf` file has changed, in which case it re-converts
 
 ---
 
 **EDIT: I've made the filter into an extension which can be found here: https://github.com/roaldarbol/pdf-to-image**
 
 ---

```lua
-- Lua filter for Quarto to convert PDF images to WebP only in HTML output

-- Check if the current output format is HTML
local is_html = (quarto and quarto.doc.is_format("html"))

-- Function to check if a file exists
local function file_exists(name)
  local f = io.open(name, "r")
  if f then f:close() end
  return f ~= nil
end

-- Function to get the modification time of a file using a shell command
local function file_mod_time(name)
  local handle = io.popen('stat -f "%m" "' .. name .. '"') -- macOS or BSD
  local result = handle:read("*a")
  handle:close()
  
  return tonumber(result)
end

function Image(elem)
  -- Only proceed if output format is HTML and the image is a PDF
  if is_html and elem.src:match("%.pdf$") then
    -- Define new filename for the WebP version
    local webp_src = elem.src:gsub("%.pdf$", ".webp")

    -- Check if WebP file exists and compare modification times
    local pdf_time = file_mod_time(elem.src)
    local webp_time = file_exists(webp_src) and file_mod_time(webp_src)

    -- If WebP doesn't exist or the PDF is newer, convert PDF to WebP
    if not webp_time or (pdf_time and pdf_time > webp_time) then
      -- Run the ImageMagick command to convert PDF to WebP
      os.execute(string.format('magick -density 600 -colorspace RGB "%s" "%s"', elem.src, webp_src))
    end

    -- Update image source to the new or existing WebP file
    elem.src = webp_src
  end

  return elem
end

```

And added it in `_quarto.yml`:
```yml
format:
  html:
    filters:
      - pdf-to-webp.lua
```

**Reply by mcanouil** [reply]
Thanks for sharing your filter.

**Reply by roaldarbol** [reply]
Thanks for the pointers! Fun trying something new! 😄

**Reply by bastienchassagnol** [reply]
Could this feature, @roaldarbol and @mcanouil, could it be considered integrated within quarto, as a quarto extension? Since there ar emore and more image files avalaible on the web as a `webp` extension?

**Reply by mcanouil** [reply]
I can'ttell.
But this filter could at least be shared as a Quarto extension at first.

You can see extensions as some kind of incubator. If an extension is not used, integrating it makes not much sense.

**Reply by bastienchassagnol** [reply]
Actually, that was the exact reciprocal problem I had, namely converting `webp` images to `pdf` or `png` (any recommendations on the favoured format), when rendering pdf or docx outputs. I dealt (partially) with that issue also creating my own (local only) extension for converting webp images using `magick convert`:

```lua
function Image(img)
  if img.src:match("%.webp$") then
    local png_src = img.src:gsub("%.webp$", ".png")
    os.execute("magick convert " .. img.src .. " " .. png_src)
    img.src = png_src
  end
  return img
end
```

Could I consider tailor and integrate this code to yours, @roaldarbol, in order to create a more generic quarto extension for automatically handling distinct kinds of image formats depending on the output format?

**Reply by mcanouil** [reply]
Be aware that you script might not work across OS, so you should be sure your filter handle this well.

Note that the author edited his post to show that the filter is available as a Quarto extension: https://github.com/roaldarbol/pdf-to-image (which you can also find on https://m.canouil.dev/quarto-extensions/)

**Reply by roaldarbol** [reply]
Hi @bastienchassagnol!
As @mcanouil mentioned, the filter is already available as an extension. I won't have time to develop it into a more general extension - however, feel free to play around with it and make a PR if you want! :-) You're also free to make make a new extension out of it if that is easier. :-)
